File: src/js/src/events/DispImpl.js
/**
 * 事件分发基类
 * @type {[type]}
 * add by movinliao
 */
var IDispatcher = require('./IDispatcher');
var EventCenter = require('./EventCenter');
var BaseEvent = require('../base/BaseEvent');
var Handler = require('./Handler');

var DispKey = 0;

var DispImpl = IDispatcher.extend(function () {
  var self = this;
  var basekey = '';

  this.constructor = function (key) {
    ++DispKey;
    basekey = DispKey + '.' + key;
  };

  this.addListener = function (key, handler, priority) {
    EventCenter.addListener(basekey, key, self, new Handler(null, handler, self), priority);
  };

  this.removeListener = function (key, handler, priority) {
    EventCenter.removeListener(basekey, key, self, new Handler(null, handler, self), priority);
  };

  this.dispatch = function (key, event) {
    EventCenter.dispatch(basekey, key, event);
  };

  this.doEvent = function (key, param) {
    var event = new BaseEvent(key);
    event.eventData = param;

    self.dispatch(key, event);
  };

  this.clear = function () {
    EventCenter.clear(basekey, self);
  };

  this.setPaused = function (enabled) {
    EventCenter.setPause(basekey, enabled);
  };

  this.on = function (key, listener, handler, priority) {
    EventCenter.addListener(basekey, key, self,
      new Handler(listener, handler, listener), priority);
  };

  this.off = function (key, listener, handler, priority) {
    EventCenter.removeListener(basekey, key, self,
      new Handler(listener, handler, listener), priority);
  };
});

module.exports = DispImpl;

File: src/c++/event/dispimpl.cpp
#include "stable.h"

#include "dispimpl.h"
#include "common/event/impl/eventcenter.h"
#include "common/event/handler.h"

namespace event{
	DisImpl::DisImpl(QObject* parent)
		: QObject(parent)
	{
		_baseKey = QString("%1>%2>").arg(QString::fromStdString(this->metaObject()->className())).arg(quint32(this)).toStdWString();
	}

	DisImpl::~DisImpl(void)
	{
	}

	void DisImpl::addListener( const std::wstring& key, const char * member, quint32 priority )
	{
		EventCenter::addListener(_baseKey, key, this, HandlerPtr(new Handler(this, member)), priority);
	}

	void DisImpl::removeListener( const std::wstring& key, const char * member, quint32 priority )
	{
		EventCenter::removeListener(_baseKey, key, this, HandlerPtr(new Handler(this, member)), priority);
	}

	void DisImpl::dispatch( const std::wstring& key, BaseEventPtr evt )
	{
		EventCenter::dispatch(_baseKey, key, evt);
	}

	void DisImpl::doEvent( const std::wstring& key, ParamPtr param )
	{
		BaseEventPtr event(new BaseEvent(key, param));
		dispatch(key, event);
	}

	void DisImpl::clear()
	{
		EventCenter::clear(_baseKey);
	}

	void DisImpl::on( const std::wstring& key, const char * member, quint32 priority )
	{
		addListener(key, member, priority);
	}

	void DisImpl::off( const std::wstring& key, const char * member, quint32 priority )
	{
		removeListener(key, member, priority);
	}

}

